appender: Add fallback to file creation date

kaffarell opened the pull request:
When using the linux-musl target for rust, the file creation time cannot be retrieved, as the current version does not support it yet ( This will be fixed with [^0]). In the meantime, we parse the datetime from the filename and use that as a fallback.

Fixes: #2999

[^0]: https://github.com/rust-lang/rust/pull/125692

zerolfx commented:
You may need to handle the prefix and suffix before parsing the date (a reverse function of `join_date`).

https://github.com/tokio-rs/tracing/blob/c6bedbe2725830c1e78cbcdb9168de69c98e42fc/tracing-appender/src/rolling.rs#L551-L569

kaffarell commented:
Oh, yeah, you're right. Thanks for the heads up, updated it now. Extracted the the datetime from the prefix/suffix, then converted to PrimitiveDateTime (because OffsetDateTime also takes a offset), then just assume it's utc (this is fine, because we also do this when writing AFAIK).

zerolfx commented:
The prefix or suffix can also contain a dot.

kaffarell commented:
Damn, forgot about that. Should work now!

zerolfx commented on tracing-appender/src/rolling.rs:
```diff
@@ -602,7 +602,28 @@ impl Inner {
+                        (&self.log_filename_prefix, &self.log_filename_suffix)
+                    {
+                        let date = filename
+                            .split(prefix)
```
The prefix can be a substring of the suffix.

kaffarell replied:
Hmm, correct. I think it's better if remove the complete prefix then suffix from the string.

zerolfx commented on tracing-appender/src/rolling.rs:
```diff
@@ -602,7 +602,28 @@ impl Inner {
+                            .ok()?
+                            .assume_utc()
+                    } else {
+                        PrimitiveDateTime::parse(filename, &self.date_format)
```
What if the prefix is configured but the suffix is not?

kaffarell replied:
Oops, forgot that branch.

kaffarell commented:
Third time's the charm :)
LMK what you think now!

zerolfx commented:
`replacen` can occur within a string. We must ensure they are strictly prefixes or suffixes.

kaffarell commented:
What do you mean exactly, could you paste an example? If the logfile prefix given is not actually the prefix in the filename, we will get a wrong date string and the parsing will fail, but this is intended.

mladedav commented:
I think you could have suffix `2024` and  then files like `prefix-2024-01-01-2024` will get turned into `-01-01-2024`.

kaffarell commented:
Oof, right. I could reverse the suffix and the whole filename string and then search through? Like that I would always get the first substring from behind. AFAIK rust doesn't have a function to make it more convention ...

mladedav commented:
Wouldn't `strip_prefix` and `strip_suffix` work instead of `replacen`?

bnjbvr commented:
Hi! what's the status of this PR? Looks like it addressed all the review comments, is it getting blocked for some reason? We'd love to have this fix to avoid using a fork of `tracing` that includes it, in a downstream project of ours :pray: Thanks for your work here!

kaffarell commented:
@davidbarsky friendly ping :)

tpoliaw commented on tracing-appender/src/rolling.rs:
```diff
@@ -602,7 +602,19 @@ impl Inner {
+                let created = metadata.created().ok().or_else(|| {
+                    let mut datetime = filename;
+                    if let Some(prefix) = &self.log_filename_prefix {
+                        datetime = datetime.strip_prefix(&format!("{prefix}."))?;
```
This might be an unnecessary optimisation or one that the compiler does anyway but would it make sense to split these strip methods to save building two new strings for every file in a potentially full directory?

```suggestion
                        datetime = datetime.strip_prefix(prefix)?;
                        datetime = datetime.strip_prefix('.')?;
```

kaffarell replied:
nope, doesn't get optimized away, LGTM!

tpoliaw reviewed (approved):
This looks like it'd be a fix for our use case and save us having to implement our own file rolling. Is there anything else we can do to help get this merged?

cyt-666 commented:
not work when rotation is set to Rotation::DAILY, PrimitiveDateTime::parse(datetime, &self.date_format) return None.

jplatte reviewed (approved):
Seems sensible.

jplatte commented:
Test failure unrelated (also happens on `main`).
